lib/spack/spack/fetch_strategy.py

```diff
@@ -289,24 +325,28 @@ def fetch(self):
         if save_file:
             os.rename(partial_file, save_file)
 
-        if not self.archive_file:
+        if not os.path.exists(self.archive_file):
             raise FailedDownloadError(self.url)
 
-    @property
-    def archive_file(self):
-        """Path to the source archive within this stage directory."""
-        return self.stage.archive_file
+        # Checksum validation
+        if validate:
+            self._check()
+
+        # Archive expansion
+        et = has_expanded_tree(self.source_dir)
+        do_expansion = expanded_source_tree in (any, True)
+        if do_expansion and not et:
```

Comment:
I find this whole if statement hard to follow -- also "do_expansion" reads awkwardly to me -- `expand` is a verb so why not just `expand`?

I would simplify it this way:

```python
if expanded_source_tree in (any, True):
    if has_expanded_tree(self.source_dir):
        tty.msg('Already expanded...')
    else:
        self._expand_archive(self.source_dir)
```

Now there are only two conditions to consider and not 2 variables + 2 cases with `and`/`and not`